src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+        # those specifiers depend on become the in-edges of each vertex. The specifiers
+        # are then sorted topologically according to this graph.
+        order = []
+        for spec in specifiers:
+            spec._dfs_state = 0
+
+        def dfs(spec):
+            if spec._dfs_state == 2:    # finished processing this specifier
+                return
+            elif spec._dfs_state == 1:  # specifier is being processed
+                raise RuntimeParseError(f'specifier {spec.name} '
+                                        'depends on itself')
+            spec._dfs_state = 1
+
+            # Recurse on dependencies
+            for dep in spec.requiredProperties:
+                child = properties.get(dep)
+                if child is None:
+                    raise RuntimeParseError(f'property {dep} required by '
+                                            f'specifier {spec} is not specified')
+                else:
+                    dfs(child)
+
+            # If this is a modifying specifier, recurse on the specifier
+            # that specifies the property being modified.
+            if spec in modifying_inv:
+                specifying_spec = properties[modifying_inv[spec]]
+                dfs(specifying_spec)
+
+            order.append(spec)
+            spec._dfs_state = 2
+
+        for spec in specifiers:
+            dfs(spec)
+        assert len(order) == len(specifiers)
+        for spec in specifiers:
+            del spec._dfs_state
+
+        # Establish a boolean array tracking which properties will be modified.
+        self._mod_tracker = {prop: True for prop in modifying}
+
+        # Evaluate and apply specifiers, using actual_props to indicate which properties
+        # it should actually specify.
+        self.properties = []        # will be filled by calls to _specify below
+        self._propertiesSet = set()
+        self._evaluated = DefaultIdentityDict()     # temporary cache for lazily-evaluated values
+        for spec in order:
+            spec.applyTo(self, actual_props[spec], overriding=overriding)
+        del self._evaluated
+        self.properties = tuple(sorted(self.properties))
+        assert self._propertiesSet == set(properties)
+        self._constProps = frozenset({
+            prop for prop in _defaultedProperties
+            if not needsSampling(getattr(self, prop))
+        })
+
+        # Check that all modifications have been applied and then delete tracker
+        assert not any(self._mod_tracker.values())
+        del self._mod_tracker
+
+    def _specify(self, prop, value, overriding=False):
+        # If this property has already been set, perform additional checks.
+        if prop in self._propertiesSet and not overriding:
+            if prop not in self._mod_tracker:
+                # This prop is not being modified, so ensure we aren't resetting it.
+                assert prop not in self.properties, f"Resetting (not modifying) {prop}"
```

Comment:
As far as I can tell, it's impossible for `self.properties` and `self._propertiesSet` to have different elements, so this assertion would always fail if reached. However, I also think this line is unreachable! If the intent is to catch double-specification of non-modified properties, I think this should just be `assert False`. Am I missing something?